### PR TITLE
Decompile text printer code

### PR DESCRIPTION
**Pull request by tgsm**

### CODE REVIEW

**Comment by red031000** [comment]
only thing I've gotta say is to make sure this syncs up with diamond, since that's already done there

**Comment by red031000** [comment]
https://github.com/pret/pokediamond/blob/master/arm9/src/text.c

File: src/text.c [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    void *charData = GfGfxLoader_GetCharData(NARC_graphic_font, 6, FALSE, &g2dCharData, HEAP_ID_DEFAULT);
+    MI_CpuCopy32(g2dCharData->pRawData, ret, 32 * 24 * sizeof(u16));
+    FreeToHeap(charData);
+
+    return ret;
+}
+
+void sub_020204FC(TextPrinter *printer, u32 x, u32 y, u16 fieldNum) {
+    (void)x;
+    (void)y;
```
Comment (cbt6) [inline src/text.c:296]:
```suggestion
void sub_020204FC(TextPrinter *printer, u32 unusedX, u32 unusedY, u16 fieldNum) {
```

File: src/text.c [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+            case 1:
+                DestroyTextPrinterSysTask(printer->id);
+                return;
+        }
+    } else {
+        printer->unk2D = printer->callback(&printer->template, printer->unk2E);
+    }
+}
+
+static u32 sub_02020358(TextPrinter *printer) {
```
Comment (PikalaxALT) [inline src/text.c:227]:
```suggestion
static u32 RenderFont(TextPrinter *printer) {
```

File: src/text.c [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+        return;
+    }
+
+    if (printer->unk2D == 0) {
+        printer->unk2E = 0;
+
+        GenerateFontHalfRowLookupTable(printer->template.fgColor, printer->template.bgColor, printer->template.shadowColor);
+
+        switch (sub_02020358(printer)) {
+            case 0:
```
Comment (PikalaxALT) [inline src/text.c:210]:
```suggestion
            case RENDER_PRINT:
```

File: src/text.c [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+                CopyWindowToVram(printer->template.window);
+                // fallthrough
+            case 3:
```
Comment (PikalaxALT) [inline src/text.c:213]:
```suggestion
            case RENDER_UPDATE:
```

File: src/text.c [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+                if (printer->callback != NULL) {
+                    printer->unk2D = printer->callback(&printer->template, printer->unk2E);
+                }
+                return;
+            case 1:
```
Comment (PikalaxALT) [inline src/text.c:218]:
```suggestion
            case RENDER_FINISH:
```

File: src/text.c [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+    } else {
+        printer->unk2D = printer->callback(&printer->template, printer->unk2E);
+    }
+}
+
+static u32 sub_02020358(TextPrinter *printer) {
+    u32 ret;
+    do {
+        ret = FontID_RenderText(printer->template.fontId, printer);
+    } while (ret == 2);
```
Comment (PikalaxALT) [inline src/text.c:231]:
```suggestion
    } while (ret == RENDER_REPEAT);
```

File: src/text.c [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+    dest16[13] = sFontHalfRowLookupTable[(u32)src16[6] & 0xFF];
+    dest16[14] = sFontHalfRowLookupTable[(u32)src16[7] >> 8];
+    dest16[15] = sFontHalfRowLookupTable[(u32)src16[7] & 0xFF];
+}
+
+static void sub_020204B8(TextPrinter *printer) {
+    printer->unk30 = NULL;
+}
+
+static u16 *sub_020204C0(void) {
```
Comment (PikalaxALT) [inline src/text.c:283]:
```suggestion
static u16 *LoadScreenFocusIndicatorGraphics(void) {
```

File: src/text.c [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+
+    NNSG2dCharacterData *g2dCharData;
+    void *charData = GfGfxLoader_GetCharData(NARC_graphic_font, 6, FALSE, &g2dCharData, HEAP_ID_DEFAULT);
+    MI_CpuCopy32(g2dCharData->pRawData, ret, 32 * 24 * sizeof(u16));
+    FreeToHeap(charData);
+
+    return ret;
+}
+
+void sub_020204FC(TextPrinter *printer, u32 x, u32 y, u16 fieldNum) {
```
Comment (PikalaxALT) [inline src/text.c:294]:
```suggestion
void RenderScreenFocusIndicatorTile(TextPrinter *printer, u32 x, u32 y, u16 fieldNum) {
```

**Review by PikalaxALT — CHANGES_REQUESTED** [review]

**Comment by tgsm** [comment]
I'll merge this soon if there are no more issues

**Review by PikalaxALT — APPROVED** [review]

**Comment by red031000** [comment]
hold on gimmie a chance to have a look at it, will try get to it tonight

File: include/font.h [hunk 9]
```diff
@@ -11,7 +11,7 @@ void FontID_SetAccessDirect(FontID fontId, HeapID heapId);
 void FontID_SetAccessLazy(FontID fontId);
 void FontID_Release(FontID fontId);
 struct GlyphInfo *FontID_TryLoadGlyph(FontID fontId, u16 glyphId);
-BOOL FontID_RenderText(int fontId, struct TextPrinter *printer);
+enum RenderResult FontID_RenderText(int fontId, struct TextPrinter *printer);
```
Comment (red031000) [inline include/font.h:14]: can't remember if we agreed to typedef enums, if so, no enum keyword here

File: include/font_types_def.h [hunk 10]
```diff
@@ -57,10 +57,17 @@ struct TextPrinter {
...
-    u8 Unk2D;
-    u16 Unk2E;
-    u16 *Unk30;
+    u8 id;
+    u8 unk2D;
+    u16 unk2E;
+    u16 *unk30;
+} TextPrinter;
+
+enum RenderResult {
```
Comment (red031000) [inline include/font_types_def.h:66]: might need to typedef this one

File: include/render_text.h [hunk 11]
```diff
@@ -13,7 +13,7 @@ struct TextPrinterSubStruct {
     u8 autoScrollDelay : 8;
 };
 
-BOOL RenderText(struct TextPrinter *printer);
+enum RenderResult RenderText(struct TextPrinter *printer);
```
Comment (red031000) [inline include/render_text.h:16]: maybe no enum keyword

File: src/alph_puzzle.c [hunk 12]
```diff
@@ -823,7 +823,7 @@ void ov110_021E6904(AlphPuzzleData *data, int a1) {
     }
     
     u32 width = FontID_String_GetWidth(4, data->quitText, 0);
-    AddTextPrinterParameterized2(data->window, 4, data->quitText, (48 - width) / 2, 4, 0xFF, ov110_021E6DA4[a1], NULL);
+    AddTextPrinterParameterizedWithColor(data->window, 4, data->quitText, (48 - width) / 2, 4, 0xFF, ov110_021E6DA4[a1], NULL);
```
Comment (red031000) [inline src/alph_puzzle.c:826]: may be out of scope, but ensure that the colours are generated using the appropriate macro here (that may have been my job that I forgot to do in my last PR, whoops)

Comment (red031000) [reply src/alph_puzzle.c:826]: and text speed constants

File: src/credits/credits.c [hunk 13]
```diff
@@ -248,7 +248,7 @@ BOOL CreditsApp_OvyInit(OVY_MANAGER *man, int *state) {
         // Print game title
         ReadMsgDataIntoString(work->pageWork.msgData, GAME_TITLE_MSG_NO, work->pageWork.string);
         FillWindowPixelBuffer(&work->pageWork.window, 0);
-        AddTextPrinterParameterized2(&work->pageWork.window, 0, work->pageWork.string, 0, 70, TEXT_SPEED_NOTRANSFER, 0x00010200, NULL);
+        AddTextPrinterParameterizedWithColor(&work->pageWork.window, 0, work->pageWork.string, 0, 70, TEXT_SPEED_NOTRANSFER, 0x00010200, NULL);
```
Comment (red031000) [inline src/credits/credits.c:251]: colour macro

File: src/credits/credits.c [hunk 14]
```diff
@@ -930,7 +930,7 @@ static void LoadPage(PageWork *ptr) {
             u8 xPos = sCreditMsgLines[ptr->pageState.line].x;
             u8 yPos = sCreditMsgLines[ptr->pageState.line].y;
             ReadMsgDataIntoString(ptr->msgData, msgno, ptr->string);
-            AddTextPrinterParameterized2(&ptr->window, 0, ptr->string, xPos, yPos, TEXT_SPEED_NOTRANSFER, 0x00010200, NULL);
+            AddTextPrinterParameterizedWithColor(&ptr->window, 0, ptr->string, xPos, yPos, TEXT_SPEED_NOTRANSFER, 0x00010200, NULL);
```
Comment (red031000) [inline src/credits/credits.c:933]: colour macro

File: src/font.c [hunk 15]
```diff
@@ -102,7 +102,7 @@ struct GlyphInfo *FontID_TryLoadGlyph(FontID fontId, u16 glyphId) {
     return &sFontWork->glyph_buffer;
 }
 
-BOOL FontID_RenderText(int fontId, struct TextPrinter *printer) {
+enum RenderResult FontID_RenderText(int fontId, struct TextPrinter *printer) {
```
Comment (red031000) [inline src/font.c:105]: possibly no enum here, remove struct as well

File: src/list_menu.c [hunk 16]
```diff
@@ -178,9 +178,9 @@ s32 ListMenuGetTemplateField(struct ListMenu *list, enum ListMenuAttr attr) {
 static void ListMenuPrint(struct ListMenu *list, String *str, u8 x, u8 y) {
     if (str != NULL) {
         if (list->overrideEnabled) {
-            AddTextPrinterParameterized3(list->template.window, list->fontId, str, x, y, 0xFF, MAKE_TEXT_COLOR(list->cursorPal, list->cursorShadowPal, list->fillValue), list->lettersSpacing, 0, NULL);
+            AddTextPrinterParameterizedWithColorAndSpacing(list->template.window, list->fontId, str, x, y, 0xFF, MAKE_TEXT_COLOR(list->cursorPal, list->cursorShadowPal, list->fillValue), list->lettersSpacing, 0, NULL);
```
Comment (red031000) [inline src/list_menu.c:181]: text speed constants?

File: src/list_menu_2d.c [hunk 17]
```diff
@@ -186,7 +186,7 @@ static void Print2dMenuItemTexts(struct ListMenu2D *listMenu) {
                 listMenu->template.fontId,
                 listMenu->template.items[j + i * listMenu->template.itemsHigh].text,
                 x,
-                listMenu->y + (listMenu->maxGlyphHeight + listMenu->template.yTop) * j,
+                (u8)(listMenu->y + (listMenu->maxGlyphHeight + listMenu->template.yTop) * j),
```
Comment (red031000) [inline src/list_menu_2d.c:189]: this is going to conflict with the other PR

File: src/scrcmd_c.c [hunk 18]
```diff
@@ -787,7 +787,7 @@ BOOL ScrCmd_DirectionSignpost(ScriptContext* ctx) {
     ReadMsgDataIntoString(ctx->msgdata, msg_no, *tmp_str);
     StringExpandPlaceholders(*msg_fmt, *unk1, *tmp_str);
     Window* window = ov01_021F3D80(fieldSystem->unk68);
-    AddTextPrinterParameterized2(window, 1, *unk1, 0, 0, 0, MAKE_TEXT_COLOR(2, 10, 15), NULL);
+    AddTextPrinterParameterizedWithColor(window, 1, *unk1, 0, 0, 0, MAKE_TEXT_COLOR(2, 10, 15), NULL);
```
Comment (red031000) [inline src/scrcmd_c.c:790]: text speed

File: src/text.c [hunk 19]
```diff
@@ -0,0 +1,314 @@
+#include "global.h"
+#include "font.h"
+#include "gf_gfx_loader.h"
+#include "text.h"
+#include "unk_0201F79C.h"
+
+static const struct FontInfo *gFonts;
```
Comment (red031000) [inline src/text.c:7]: sFonts

Comment (red031000) [reply src/text.c:7]: and no struct

File: src/text.c [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+static u8 _021D1F6C;
+static u16 sFgColor, sShadowColor, sBgColor;
+static SysTask *sTextPrinterTasks[MAX_TEXT_PRINTERS];
+static u16 sFontHalfRowLookupTable[4 * 4 * 4 * 4];
+
+static u8 CreateTextPrinterSysTask(SysTaskFunc taskFunc, TextPrinter *printer, u32 priority);
+static BOOL TextPrinterSysTaskIsActive(u8 printerId);
+static u8 AddTextPrinter(TextPrinterTemplate *template, u32 speed, PrinterCallback_t callback);
+static void RunTextPrinter(SysTask *task, TextPrinter *printer);
+static enum RenderResult RenderFont(TextPrinter *printer);
```
Comment (red031000) [inline src/text.c:18]: possibly no enum here

File: src/text.c [hunk 21]
```diff
@@ -0,0 +1,314 @@
...
+static void sub_020204B8(TextPrinter *printer);
+static u16 *LoadScreenFocusIndicatorGraphics(void);
+static void sub_02020548(TextPrinter *printer);
+
+void SetFontsPointer(const struct FontInfo *fonts) {
```
Comment (red031000) [inline src/text.c:23]: no struct

File: src/text.c (hunk 19 above)
Comment (red031000) [reply src/text.c:7]: (I understand this is different in diamond but it will be updated there as well to be correct)

File: src/text.c [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+    return AddTextPrinter(&template, textSpeed, callback);
+}
+
+static u8 AddTextPrinter(TextPrinterTemplate *template, u32 speed, PrinterCallback_t callback) {
+    if (gFonts == NULL) {
+        return 0xFF;
+    }
+
+    TextPrinter *printer = AllocFromHeap(HEAP_ID_DEFAULT, sizeof(TextPrinter));
+    printer->active = 1;
```
Comment (red031000) [inline src/text.c:153]: TRUE

File: src/text.c [hunk 23]
```diff
@@ -0,0 +1,314 @@
...
+            case RENDER_FINISH:
+                DestroyTextPrinterSysTask(printer->id);
+                return;
+        }
+    } else {
+        printer->unk2D = printer->callback(&printer->template, printer->unk2E);
+    }
+}
+
+static enum RenderResult RenderFont(TextPrinter *printer) {
```
Comment (red031000) [inline src/text.c:227]: possibly no enum

File: src/voltorb_flip/voltorb_flip.c [hunk 24]
```diff
@@ -1533,7 +1533,7 @@ static void PrintMessageOnWindow(VoltorbFlipAppWork *work, FontID fontId, u8 msg
     FillWindowPixelBuffer(window, 0);
 
     String *str = ReadMsgData_ExpandPlaceholders(work->msgFmt, work->msgData, msgNo, work->heapId);
-    AddTextPrinterParameterized2(window, fontId, str, x, y, 0xff, textSpeed, 0);
+    AddTextPrinterParameterizedWithColor(window, fontId, str, x, y, 0xff, textSpeed, 0);
```
Comment (red031000) [inline src/voltorb_flip/voltorb_flip.c:1536]: text speed is mis-labeleld here, this is colour, the 0xff should be the constant

**Review by red031000 — CHANGES_REQUESTED** [review]
mostly good, apart from a few small things, some may be out of scope depending on how much cleanup you want to do

File: src/text.c (hunk 19 above)
Comment (tgsm) [reply src/text.c:7]: Losing `struct` can be done separately

File: src/text.c (hunk 21 above)
Comment (tgsm) [reply src/text.c:23]: Can be resolved separately

File: src/text.c (hunk 19 above)
Comment (red031000) [reply src/text.c:7]: it *can* be, but it *shouldn't* be imo

**Review by red031000 — APPROVED** [review]
could probably use more constants for arbitrary values, but satisfied for noe

**Pull request merged**